src/motor_node.cc

```diff
@@ -53,49 +57,146 @@ void PID_update_callback(const ubiquity_motor::PIDConfig& config,
+        } else {
+            ROS_WARN("Invalid wheel_direction of '%s' specified! Using wheel direction of standard", 
+                g_node_params.wheel_direction.c_str());
+            g_node_params.wheel_direction = "standard";
+            wheel_direction = MotorMessage::OPT_WHEEL_DIR_STANDARD;
+        }
+        // Write out the wheel direction setting
+        robot->setWheelDirection(wheel_direction);
+        mcbStatusPeriodSec.sleep();
+    }
+
+    // Tell the controller board firmware what version the hardware is at this time.
+    // TODO: Read from I2C.   At this time we only allow setting the version from ros parameters
+    if (robot->firmware_version >= MIN_FW_HW_VERSION_SET) {
+        ROS_INFO_ONCE("Firmware is version %d. Setting Controller board version to %d", 
+            robot->firmware_version, g_firmware_params.controller_board_version);
+        robot->setHardwareVersion(g_firmware_params.controller_board_version);
+        ROS_DEBUG("Controller board version has been set to %d", 
+            g_firmware_params.controller_board_version);
+        mcbStatusPeriodSec.sleep();
+    }
+
+    // Tell the MCB board what the I2C port on it is set to (mcb cannot read it's own switchs!)
+    // We could re-read periodically but perhaps only every 5-10 sec but should do it from main loop
+    if (robot->firmware_version >= MIN_FW_OPTION_SWITCH) {
+        g_firmware_params.option_switch = robot->getOptionSwitch();
+        ROS_INFO("Setting firmware option register to 0x%x.", g_firmware_params.option_switch);
+        robot->setOptionSwitchReg(g_firmware_params.option_switch);
+        mcbStatusPeriodSec.sleep();
+    }
+    
+    if (robot->firmware_version >= MIN_FW_SYSTEM_EVENTS) {
+        // Start out with zero for system events
+        robot->setSystemEvents(0);  // Clear entire system events register
+        robot->system_events = 0;
+        mcbStatusPeriodSec.sleep();
+    }
+
+    // Setup other firmware parameters that could come from ROS parameters
+    if (robot->firmware_version >= MIN_FW_ESTOP_SUPPORT) {
+        robot->setEstopPidThreshold(g_firmware_params.estop_pid_threshold);
+        mcbStatusPeriodSec.sleep();
+        robot->setEstopDetection(g_firmware_params.estop_detection);
+        mcbStatusPeriodSec.sleep();
+    }
+
+    if (robot->firmware_version >= MIN_FW_MAX_SPEED_AND_PWM) {
+        robot->setMaxFwdSpeed(g_firmware_params.max_speed_fwd);
+        mcbStatusPeriodSec.sleep();
+        robot->setMaxRevSpeed(g_firmware_params.max_speed_rev);
+        mcbStatusPeriodSec.sleep();
+    }
+        
+    return;
+}
+
 int main(int argc, char* argv[]) {
     ros::init(argc, argv, "motor_node");
     ros::NodeHandle nh;
 
-    firmware_params = FirmwareParams(nh);
-    serial_params = CommsParams(nh);
-    node_params = NodeParams(nh);
-
-    ros::Rate ctrlLoopDelay(node_params.controller_loop_rate);
+    g_firmware_params = FirmwareParams(nh);
```

Comment:
Nit: these aren't really globals, so not sure the `g_` prefix makes sense.

Reply:
Ok, got carried away.  I like to use g_ when something is a var in process global space and it should not be g_ if it is a member variable of any class.

Reply:
Actually I just looked.   These were globals before without g_  so I added g_
firmware_params, serial_params and node_params were all globals before but with no g_
the g_ makes it clear now that we are talking about the static  FirmwareParams in motor_node.cc

They have been globals or at least   static casts in motor_node.cc (a scoped global if you will) since 2017

Reply:
I am loosely using global here to mean not in a class.  Not totally open globals but static cast and for the sake of the code in motor_node.cc for those purposes they are global.